spring-cloud-kubernetes-fabric8-discovery/src/main/java/org/springframework/cloud/kubernetes/fabric8/discovery/KubernetesCatalogWatch.java

```diff
@@ -41,23 +40,23 @@
  */
 public class KubernetesCatalogWatch implements ApplicationEventPublisherAware {
 
-	private static final LogAccessor LOG = new LogAccessor(LogFactory.getLog(KubernetesCatalogWatch.class));
+	private static final String DISCOVERY_GROUP_VERSION = "discovery.k8s.io/v1";
+
+	private static final String ENDPOINT_SLICE = "EndpointSlice";
 
-	private final KubernetesClient kubernetesClient;
+	private static final LogAccessor LOG = new LogAccessor(LogFactory.getLog(KubernetesCatalogWatch.class));
 
-	private final KubernetesDiscoveryProperties properties;
+	private final Fabric8CatalogWatchContext context;
 
-	private final KubernetesNamespaceProvider namespaceProvider;
+	private Function<Fabric8CatalogWatchContext, List<EndpointNameAndNamespace>> stateGenerator;
```

Comment:
we separate the logic now into two logical pieces: into Endpoints and EndpointSlices. Each of those (`Fabric8EndpointsCatalogWatch` and `Fabric8EndpointSliceV1CatalogWatch`) is actually a `Function`. This function takes as input a `Fabric8CatalogWatchContext` (which has members as : `KubernetesClient`, `KubernetesDiscoveryProperties` and `KubernetesNamespaceProvider` ) and returns a `List<EndpointNameAndNamespace>`.